### PR TITLE
fix suicide and autoExec upgrade

### PR DESCRIPTION
**Pull request by ouwenkg**
### Description of the Change

Fix two issues about upgrade:

1. Suicide behavior
2. Auto exec coinbase.

### CODE REVIEW

**Comment by ouwenkg** [comment]
### 测试一 自动执行

测试合约如下:

```
pragma solidity ^0.4.24;

contract X {
    address public addr;
    uint public x;

    constructor() payable public {
        addr = block.coinbase;
    }

    function autoExec() public {
        addr = block.coinbase;
        // address(block.coinbase).transfer(1);
        x++;
    }
}

======= test/auto_exec.sol:X =======
Function signatures: 
767800de: addr()
844cbc43: autoExec()
0c55699c: x()
```

部署合约
```
$ rpc sendRawTransaction --code 0x6080604052416000806101000a81548173ffffffffffffffffffffffffffffffffffffffff021916908373ffffffffffffffffffffffffffffffffffffffff1602179055506101a0806100536000396000f300608060405260043610610057576000357c0100000000000000000000000000000000000000000000000000000000900463ffffffff1680630c55699c1461005c578063767800de14610087578063844cbc43146100de575b600080fd5b34801561006857600080fd5b506100716100f5565b6040518082815260200191505060405180910390f35b34801561009357600080fd5b5061009c6100fb565b604051808273ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff16815260200191505060405180910390f35b3480156100ea57600080fd5b506100f3610120565b005b60015481565b6000809054906101000a900473ffffffffffffffffffffffffffffffffffffffff1681565b416000806101000a81548173ffffffffffffffffffffffffffffffffffffffff021916908373ffffffffffffffffffffffffffffffffffffffff1602179055506001600081548092919060010191905055505600a165627a7a72305820e3806297d7515cccf957a397240f08edf834ce4163da12365d08bbf9134f4e0d0029 --private-key 0x5f0258a4778057a8a7d97809bd209055b2fbafa654ce7d31ec7191066b9225e6
```

合约地址: 0xe0942d7c758a9332c0e472b8ee04b5486893aa90

查询 coinbase:
```
$rpc call --to 0xe0942d7c758a9332c0e472b8ee04b5486893aa90 --data 0x767800de
{
  "id": 1,
  "jsonrpc": "2.0",
  "result": "0x000000000000000000000000e2fdb9bc00bb1a19433ddfa42647d3815d8745dc"
}

$ rpc getMetaData
{
  "id": 1,
  "jsonrpc": "2.0",
  "result": {
    "blockInterval": 3000,
    "chainId": 0,
    "chainIdV1": "0x1",
    "chainName": "test-chain",
    "economicalModel": 1,
    "genesisTimestamp": 1557385482959,
    "operator": "test-operator",
    "tokenAvatar": "https://cdn.cryptape.com/icon_cita.png",
    "tokenName": "CITA Test Token",
    "tokenSymbol": "CTT",
    "validators": [
      "0xe2fdb9bc00bb1a19433ddfa42647d3815d8745dc"
    ],
    "version": 2,
    "website": "https://www.example.com"
  }
```

可以看到自动执行里查到的 coinbase 就是出块节点.

### 测试二  Suicide 操作

测试合约:

```
pragma solidity ^0.4.24;

contract LifeTimeExample {
    address public owner;
    address public refund_account = 0x7ea9be12d3a939a96699ff5e96e1d85034a94f01;

    constructor() public {
        owner = msg.sender;
    }

    function() public payable {}

    function kill() public {
        selfdestruct(refund_account);
    }
}

Function signatures: 
41c0e1b5: kill()
8da5cb5b: owner()
8d37008c: refund_account()
```

部署合约:

```bash
$ rpc sendRawTransaction --code 0x608060405273e1c4021742730ded647590a1686d5c4bfcbae0b0600160006101000a81548173ffffffffffffffffffffffffffffffffffffffff021916908373ffffffffffffffffffffffffffffffffffffffff16021790555034801561006557600080fd5b50336000806101000a81548173ffffffffffffffffffffffffffffffffffffffff021916908373ffffffffffffffffffffffffffffffffffffffff1602179055506101d0806100b56000396000f300608060405260043610610057576000357c0100000000000000000000000000000000000000000000000000000000900463ffffffff16806341c0e1b5146100595780638d37008c146100705780638da5cb5b146100c7575b005b34801561006557600080fd5b5061006e61011e565b005b34801561007c57600080fd5b50610085610159565b604051808273ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff16815260200191505060405180910390f35b3480156100d357600080fd5b506100dc61017f565b604051808273ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff16815260200191505060405180910390f35b600160009054906101000a900473ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff16ff5b600160009054906101000a900473ffffffffffffffffffffffffffffffffffffffff1681565b6000809054906101000a900473ffffffffffffffffffffffffffffffffffffffff16815600a165627a7a7230582056c0b6f079da3e8bed7f23d8da648b7b745c5bd87a6ddcaae03774f1be4c478d0029 --private-key 0x5f0258a4778057a8a7d97809bd209055b2fbafa654ce7d31ec7191066b9225e6
```

得到地址: 0xc0572185475332eea0333e6da1a70e2f2a97def4', 先给合约账户转一笔钱.

```bash
$ transfer --address 0xc0572185475332eea0333e6da1a70e2f2a97def4 --private-key 0x5f0258a4778057a8a7d97809bd209055b2fbafa654ce7d31ec7191066b9225e6 --value 9999
```

先查询 suicide 之前的账户余额:

```
查询合约账户余额:
$ rpc getBalance --address 0xc0572185475332eea0333e6da1a70e2f2a97def4
{
  "id": 1,
  "jsonrpc": "2.0",
  "result": "0x270f"
}

查询 refund account 余额:
rpc getBalance --address 0x7ea9be12d3a939a96699ff5e96e1d85034a94f01
{
  "id": 1,
  "jsonrpc": "2.0",
  "result": "0x0"
}
```

发交易触发 suicide 操作:

```bash
$ rpc sendRawTransaction --address 0xc0572185475332eea0333e6da1a70e2f2a97def4 --code 0x41c0e1b5 --private-key 0x5f0258a4778057a8a7d97809bd209055b2fbafa654ce7d31ec7191066b9225e6
```

查询 suicide 之后的账户余额情况:

```
查询合约账户余额:
$ rpc getBalance --address 0xc0572185475332eea0333e6da1a70e2f2a97def4
{
  "id": 1,
  "jsonrpc": "2.0",
  "result": "0x0"
}


查询 refund account 余额:
$ rpc getBalance --address 0x7ea9be12d3a939a96699ff5e96e1d85034a94f01
{
  "id": 1,
  "jsonrpc": "2.0",
  "result": "0x270f"
}
```

可以看到, suicide 后, 合约账户的余额转入refund account.

**Comment by ouwenkg** [comment]
关于这两个功能点的自动化测试, 准备在升级结束后加上.

**Comment by kaikai1024** [comment]
自动执行缺少注册的步骤

File: cita-executor/core/src/externalities.rs [hunk 1]
```diff
@@ -418,6 +423,16 @@ where
 
         let address = self.origin_info.address;
         let balance = self.balance(&address)?;
+
+        if self.chain_version > 1 {
+            if &address == refund_address {
```
Comment (kaikai1024) [inline cita-executor/core/src/externalities.rs]: If refund to self, then fire the token?

**Comment by ouwenkg** [comment]
> 自动执行缺少注册的步骤

自动执行功能不用再测了吧,  出的问题在 coinbase 这里.

File: cita-executor/core/src/externalities.rs (hunk 1 above)
Comment (ouwenkg) [reply cita-executor/core/src/externalities.rs]: yes

**Comment by kaikai1024** [comment]
> 自动执行功能不用再测了吧, 出的问题在 coinbase 这里.

是自动执行的时候的 coinbase ，你那个测试本身就是没问题的。
自动执行时候获取 coinbase 才会失败

File: cita-executor/core/src/externalities.rs [hunk 2]
```diff
@@ -418,6 +423,16 @@ where
 
         let address = self.origin_info.address;
         let balance = self.balance(&address)?;
+
+        if self.chain_version > 1 {
```
Comment (leeyr338) [inline cita-executor/core/src/externalities.rs]: It is better  to define some `MACRO` for the `version` number, and then comment the difference between each version. I think they will be more readable!

Comment (ouwenkg) [reply cita-executor/core/src/externalities.rs]: In term of protocol version, we had a explanation docs to tell difference between each versions.

As for `Use Macro`, we can discuss details later. Sadly, i am not skilled to macro usage in rust.

https://docs.citahub.com/zh-CN/cita/protocol-upgrade/v0
https://docs.citahub.com/zh-CN/cita/protocol-upgrade/v1

Comment (kaikai1024) [reply cita-executor/core/src/externalities.rs]: #565

Comment (kaikai1024) [reply cita-executor/core/src/externalities.rs]: > In term of protocol version, we had a explanation docs to tell difference between each versions.

The doc and the comment two things.
Will do it at a new pr.

**Pull request merged**